modules/standard/Socket.chpl

```diff
@@ -0,0 +1,481 @@
+/*
+ * Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+module Socket {
+
+  public use Sys;
+  use Time;
+  use SysCTypes;
+  use SysError;
+  use SysBasic;
+  use CPtr;
+  use IO;
+
+  enum IPFamily {
+    IPv4 = 2,
+    IPv6 = 10,
+    IPUnspec = 0
+  }
+
+  const IPv4Localhost = INADDR_LOOPBACK;
+  const IPv6Localhost = in6addr_loopback;
+
+  record ipAddr {
+    var _addressStorage:sys_sockaddr_t;
+  }
+
+  proc ipAddr.init(inout host:string = "localhost", port:uint(16) = 8000, family:IPFamily = IPFamily.IPv4) {
+    if(host == "localhost"){
+      host = if family == IPFamily.IPv4 then "127.0.0.1" else "::1";
+    }
+
+    _addressStorage = new sys_sockaddr_t();
+
+    try! {
+      _addressStorage.set(host.c_str(), port, family:c_int);
+    }
+  }
+
+  proc ipAddr.init(host:sys_in_addr_t, port:uint(16) = 8000) {
+    _addressStorage = new sys_sockaddr_t();
+
+    try! {
+      _addressStorage.set(host, port);
+    }
+  }
+
+  proc ipAddr.init(host:sys_in6_addr_t, port:uint(16) = 8000) {
+    _addressStorage = new sys_sockaddr_t();
+
+    try! {
+      _addressStorage.set(host, port);
+    }
+  }
+
+  pragma "no doc"
+  proc ipAddr.init(in address:sys_sockaddr_t) {
+    this._addressStorage = new sys_sockaddr_t();
+    try! {
+      _addressStorage.set(address.numericHost().c_str(), address.port(), address.family);
+    }
+  }
```

Comment:
The use of this would for inside Module use mostly as to copy an ipAddr or create one we essentially need to copy the `_addressStorage` inside it which is returned from system calls